### ISSUE TITLE
Better support for repeating section

### ISSUE THREAD

**Issue by chrismosby** [formly-js/angular-formly#282]
For repeating sections, I think **ng-repeating** in templates defeat the purpose of the library. The core principle of **angular-formly** is "*configure your forms within your controller*", so I think there should be a "**multiple**" key in the configuration object that would automatically create a repeating section, by adding a plus and delete buttons. 

Also the only part of the doc that shows an example of repeating sections says that 

***"NOTE: You normally shouldn't use the formly-field directive"***
***"See the documentation for things to consider: http://docs.angular-formly.com/docs/formly-field"***

So basically no clear direction on how to handle that case.

**Comment by kentcdodds** [comment]
I updated the example to use the new `formGroup` and `className` apis. These are only available in the latest (beta) release which should hopefully be released in the next day or two. Let me know what you think.

**Comment by chrismosby** [comment]
Hey @kentcdodds it looks great! Thank you for your responsiveness. But to eliminate any kind of manual action besides defining types, is it possible to make the "ng-repeat" work under the hood ? Thank you

**Comment by kentcdodds** [comment]
I'm not sure I understand what you mean... `ng-repeat` *is* working under the hood...

**Comment by chrismosby** [comment]
ok, basically. See this code ?
```
<div>
  <!--loop through each element in model array-->
  <div class="{{hideRepeat}}">
    <div ng-repeat="element in model[options.key]">
      <formly-form
           fields="to.fields"
           model="element"
           form-state="formOptions">
      </formly-form>
      <div style="margin-bottom:20px;">
        <button type="button" class="btn btn-sm btn-danger" ng-click="model[options.key].splice($index, 1)">
          Remove
        </button>
      </div>
      <hr>
  </div>
  <p class="AddNewButton">
    <button type="button" class="btn btn-primary" ng-click="addNew()" >{{to.btnText}}</button>
  </p>
```
I don't think very useful from a user perspective. People would be just using the same code in their app. So why not just create 2 new keys in the configuration object : "Delete" and "Add new", since it's mostly the only thing that would change from user to user.

**Comment by kentcdodds** [comment]
I think all you're saying is that you'd like a pre-defined type that does this. You don't want to have to define it yourself. Is that right?

**Comment by chrismosby** [comment]
Yes!

**Comment by kentcdodds** [comment]
I see... Well, types are a template concern. And templates are specific to css frameworks. Are you looking for a bootstrap template then?

**Comment by kentcdodds** [comment]
As a side, I highly recommend that if you're going to be serious about creating forms in your app, you should create your own types. As awesome as the template libraries are, they can't beat the customizability required for a serious application. And types are so easy to create that you shouldn't have a problem with it.

**Comment by chrismosby** [comment]
I agree. I believed it wasn't necessarily template specific. We've created our custom implementation of recursive forms in my company, and we abstracted repeated sections from templates. So all you do is defining what needs to be repeated and you tell the constructor that it's repeatable and it does the work. 

something like : 

```
Form({
    fields: {
       name: {
          template: '<input>'
       },
    },
    options: {
      repeat: true,
      ...
    },
})
```

**Comment by kentcdodds** [comment]
I see. I think having that built-into angular-formly could be useful... I'll see if I can think of a good way to include that.

**Comment by chrismosby** [comment]
Awesome! I'll try to see how I can help this weekend.

**Comment by jlowery457** [comment]
What about expressionProperties for the repeating section example?  If I modify the repeating section example to add an expression property for investmentName:

`expressionProperties: {
                      'templateOptions.disabled': function ($viewValue, $modelValue, scope) {
                  return scope.model.investmentValue === '';
                }`

The investment name field in all sections is disabled even though only one of the investmentValue fields in only one of the sections satisfies the logic check.

**Comment by kentcdodds** [comment]
@jlowery457, I improved the demo: http://angular-formly.com/#/example/advanced/repeating-section

**Comment by jlowery457** [comment]
Thanks @kentcdodds!  That really helps.

**Comment by kentcdodds** [comment]
If anyone wants to discuss a good API as well as a good implementation, I would love to talk about it. I'd really love to improve support for repeating fields/fieldGroups, but I'm not certain as to the best way to go about implementing it.

**Comment by a-laughlin** [comment]
Recently started looking at formly after it popped up on angular weekly.  Fantastic effort on demos and api docs!

What about pipelines to start? That would enable a lot of flexibility with a simple imperative API that could add custom handlers for properties like 'repeat'.


So in the core, something like:
```
var pipelinesToView = {
  presentationSchema:[],// converts presentation schema (i.e., fields) into a custom object we can compile with the html
  presentationInstance:[], // operates on the output of presentation schema (for one-off change and performance);
};

// same principle from view to model, including validation, data cleanup, etc.
var pipelinesToModel = {}
```

That would also enable plugins to easily add/remove/override parts of the pipes

for example:
```
formly.addPlugin('repeater',function(pipelines){
  pipelines.presentationSchema.push(
    function repeater(presentationInstance){
      // recursively walks over each part of the presentation schema as "this"
      for(var i=0; i < this.repeats; i++){
        presentationInstance.children.push(clone(this.children[0]))
      }
      return this;
    }
  );
});

Could go with html and controller like:
```
// html
<formly-form  model="model"
              presentation-schema="presentationSchema"
              presentation-instance="presentationInstance">
</formly-form>
// controller
$scope.model ={};
$scope.presentationInstance = {};
$scope.presentationSchema = {
  fields:[{
    key:'customField',
    type:'text',
    placeholder:'foo',
    repeats:2  // <--- repeater would operate on this
  },
  // dynamically appending pipeline functions would enable sophisticated sorting/conditions/repeating, etc
  // though we'd have to think of a good way to manage conflicts
  {
    type:'button',
    events:{
      click:function(event,presentationInstance,pipelines){
        if(presentationInstance.children > 2){
          pipelines.presentationSchema.push(function(){
            // add some conditional display logic
          });
          queuePipelineExecution();
        }
      }
    }
  }
  ]
};
```

The same architecture would make it possible to implement something like the json schema in #147 by prepending transformations to the pipelines


Anyway.  Getting late.  Thanks again for your work on Formly.  Hopefully this sparks some ideas!

**Comment by kentcdodds** [comment]
I recently added the functionality to specify a `fieldTransform` on a global/form level that I think takes care of the json schema issue and may cover this case as well. Have you had a chance to look at that?

**Comment by kentcdodds** [comment]
[Docs](http://docs.angular-formly.com/v6.16.0/docs/formly-form#section-fieldtransform)

**Comment by a-laughlin** [comment]
From reading the doc, it looks like it has potential.  Examples would be fantastic.  I'll have more time to experiment with it this week.

One thing in the existing docs confuses me.  "Allows you to modify/convert the fields before angular-formly handles them or validates them."  What does "handles them" mean?  It sounds like this function would run both pre-formatters and pre-parsers/validators.

**Comment by kentcdodds** [comment]
Yes, this runs as one of the first things in the `formly-form` controller, so you can add/remove/modify any fields. Meaning what you send it can be total nonsense and as long as what you return is an array of valid field configurations, then you're good. I'll try to create an example as soon as I am able.

**Comment by kentcdodds** [comment]
Is anyone still waiting for this feature? I'm sorry I haven't really looked into implementing it. It's not the most straightforward one to implement.

**Comment by a-laughlin** [comment]
Yes.  I added basic support for repeating models in a [wrapper directive](https://github.com/formly-js/angular-formly-simplified/blob/master/angular-formly-simplified.js#L269) around formly in the angular-formly-simplified repo.  Nested repeating models are more complex, so I didn't implement those.

The basic premise is that model is always an array, and if it isn't, make it one.  That's easier to reason about than array vs. object.

I also started modifying the getFields function for preFields and postFields arrays, to add buttons that appear before and after the repeating models.  WIP.

**Comment by kentcdodds** [comment]
I'm going to close this in favor of someone creating [a plugin](http://docs.angular-formly.com/page/plugins).

**Comment by kalpitpandit** [comment]
Is there any way to call controller method on click of remove or add new button?

**Comment by ppalreddy** [comment]
Is there any way to display the remove button on the same line as form fields ?? Right now my form looks like this 
![image](https://user-images.githubusercontent.com/10407989/26989278-769d26ce-4d20-11e7-8dc2-6d79b0d94ca4.png)

I want the remove button to disaply next to registration number text box.
